preprocessing/shift_generation.py

```diff
@@ -136,14 +131,41 @@ def get_shifts_violating_daily_rest(root, staff, shifts_per_day):
                                     except:
                                         violating_shifts[e][shift] = [s]
 
+    return violating_shifts
+
+
+def get_invalid_shifts(root, staff, shifts_per_day):
```

Comment:
Nice! Breaking functions into smaller pieces makes it so much easier to both understand and work with them! 👏